ql/src/java/org/apache/hadoop/hive/ql/io/SkippingTextInputFormat.java

```diff
@@ -110,17 +110,32 @@ private long getCachedStartIndex(Path path) throws IOException {
       fileSystem = path.getFileSystem(conf);
       try {
         fis = fileSystem.open(path);
+        long currPos = fis.getPos();
+        int delimiterIdx = -1;
         for (int j = 0; j < headerCount; j++) {
-          if (fis.readLine() == null) {
+          String headerLine = fis.readLine();
+          if (headerLine == null) {
             startIndexMap.put(path, Long.MAX_VALUE);
             return Long.MAX_VALUE;
           }
+          if (j == headerCount-1) {
+            String delimiter = conf.get("textinputformat.record.delimiter");
```

Comment:
I would read the delimiter only once, outside of the for loop.

Reply:
Since the scope of delimiter is only inside the loop, just for last iteration, that's why I have defined it here. 
If it was used outside the loop as well, I would have moved it outside "for loop".